docs/develop/web-client-performance.md

```diff
@@ -0,0 +1,107 @@
+
+The most common reason for a slow script is excessive communication with the workbook. Your script runs on your local machine, while the workbook exists in the cloud. At certain times, your script synchronizes its local data with that of the workbook. This means that any write operations (such as `workbook.addWorksheet()`) are only applied to the workbook when this behind-the-scenes synchronization happens. Likewise, any read operations (such as `myRange.getValues()`) only get data from the workbook for the script at those times. In either case, the script fetches information before it acts on the data. For example, the following code will accurately log the number of rows in the used range.
+
+```TypeScript
+let usedRange = workbook.getActiveWorksheet().getUsedRange();
+let rowCount = usedRange.getRowCount();
+// The script will read the range and row count from
+// the workbook before logging the information.
+console.log(rowCount);
+```
+
+Office Scripts APIs ensure any data in the workbook or script is accurate and up-to-date when necessary. You don't need to worry about these synchronizations for your script to run correctly. However, an awareness of this script-to-cloud communication can help you avoid unneeded network calls.
+
+## Performance optimizations
+
+You can apply simple techniques to help reduce the communication to the cloud. The following patterns help speed up your scripts.
+
+- Read workbook data once instead of repeatedly in a loop.
+- Remove unnecessary `console.log` statements.
+- Avoid using try/catch blocks.
+
+### Read workbook data outside of a loop
+
+Any method that gets data from the workbook can trigger a network call. Rather than repeatedly making the same call, you should save data locally whenever possible. This is especially true when dealing with loops.
+
+Consider a script to get the count of negative numbers in the used range of a worksheet. The script needs to iterate over every cell in the used range. To do that, it needs the range, the number of rows, and the number of columns. You should store those as local variables before starting the loop. Otherwise, each iteration of the loop will force a return to the workbook.
+
+```TypeScript
+/**
+ * This script provides the count of negative numbers that are present
+ * in the used range of the current worksheet.
+ */
+function main(workbook: ExcelScript.Workbook) {
+  // Get the working range.
+  let usedRange = workbook.getActiveWorksheet().getUsedRange();
+
+  // Save the values locally to avoid repeatedly asking the workbook.
+  let usedRangeValues = usedRange.getValues();
+  let rowCount = usedRange.getRowCount();
+  let columnCount = usedRange.getColumnCount();
+
+  // Start the negative number counter.
+  let negativeCount = 0;
+
+  // Iterate over the entire range looking for negative numbers.
+  for (let i = 0; i < rowCount; i++) {
```

Comment:
This specific case could be solved with double foreach, but that's probably more complicated...

Reply:
Yeah, but I've found the fancy JS foreach loops tend to obfuscate simple examples. I'll change it if we find our customers to be more JS savvy than we expect.